src/vaadin-radio-group-mixin.ts

```diff
@@ -0,0 +1,330 @@
+  FocusVisibleInterface &
+  DisabledStateInterface &
+  FocusVisibleClass &
+  KeyboardDirectionClass;
+
+const filterRadioButtons = (nodes: Node[]) => {
+  return nodes.filter(node => {
+    return (
+      node.nodeType === Node.ELEMENT_NODE && (node.constructor as typeof RadioButtonMixinClass).hasRadioButtonMixin
+    );
+  }) as HTMLElement[];
+};
+
+let uniqueId = 0;
+
+export const RadioGroupMixin = <T extends Constructor<LitElement & KeyboardDirectionClass>>(
+  base: T
+): T & Constructor<RadioGroup> => {
+  class RadioGroup extends FocusVisibleMixin(
+    DisabledStateMixin(RovingTabIndexMixin(KeyboardDirectionMixin(SlottedItemsMixin(base))))
+  ) {
+    /**
+     * String used for the label element.
+     */
+    @property({ type: String }) label: string | null | undefined;
+
+    /**
+     * Value of the radio group.
+     */
+    @property({ type: String }) value: string | null | undefined;
+
+    /**
+     * When true, the radio group value is invalid.
+     */
+    @property({ type: Boolean, reflect: true }) invalid: boolean | null | undefined = false;
+
+    /**
+     * Radio group is aligned vertically by default. Set to "horizontal" to change this.
+     */
+    @property({ type: String, reflect: true }) orientation: string | null | undefined;
+
+    /**
+     * When true, the radio group must have a value set by the user.
+     */
+    @property({ type: Boolean }) required: boolean | null | undefined;
+
+    /**
+     * When true, the user cannot modify the value of the radio group.
+     */
+    @property({ type: Boolean }) readonly: boolean | null | undefined;
+
+    /**
+     * Error to show when the radio group value is invalid.
+     */
+    @property({ type: String, attribute: 'error-message' }) errorMessage: string | null | undefined;
+
+    private _boundCheckedChanged = this._onCheckedChanged.bind(this) as EventListener;
+
+    private _checkedButton?: RadioButton;
+
+    private _errorId = `vaadin-radio-group-error-${(uniqueId += 1)}`;
+
+    private _inputChange?: Event;
+
+    protected render() {
+      return html`
+        <div class="vaadin-group-field-container">
+          <label part="label">${this.label}</label>
+
+          <div part="group-field">
+            <slot id="slot"></slot>
+          </div>
+
+          <div
+            id="${this._errorId}"
+            part="error-message"
+            aria-live="assertive"
+            aria-hidden="${this.errorMessage && this.invalid ? 'false' : 'true'}"
+            ?hidden="${!this.invalid}"
+          >
+            ${this.errorMessage}
+          </div>
+        </div>
+      `;
+    }
+
+    constructor() {
+      super();
+
+      this.addEventListener('change', (event: Event) => {
+        if (this.items.includes(event.composedPath()[0] as HTMLElement)) {
+          event.stopImmediatePropagation();
+          // store reference to the original event
+          this._inputChange = (event as CustomEvent).detail.sourceEvent;
```

Comment:
**Not a blocker, thinking out loud:**.
As it's next gen, we can consider changing this behavior. Probably we can consider not providing the original change event from the `radio-button` and avoid problems with timings, as the `radio-button` and `radio-group` are different components and therefore should not have combined `change` events. 
`vaadin-checkbox` doesn't provide `sourceEvent` `change` either as it doesn't make sense.

Reply:
Thanks, changed. Looks like we have an inconsistency here, due to historical reasons.

### `change` with `detail.sourceEvent`

- [vaadin-text-field](https://github.com/vaadin/vaadin-text-field/blob/1bc33a2d065206177372763db2bd9ed1c5126ea0/src/vaadin-text-field-mixin.html#L417-L419) and its dependents
- [vaadin-radio-button](https://github.com/vaadin/vaadin-radio-button/blob/e2c6697b5dda7d524c0216a1e4b6795ebf22920d/src/vaadin-radio-button.html#L266-L268)

### `change` without `detail`

- [vaadin-date-picker](https://github.com/vaadin/vaadin-date-picker/blob/496b2cf6d4c109c1f3340fb58b276704c8fe7ef8/src/vaadin-date-picker-mixin.html#L536)
- [vaadin-select](https://github.com/vaadin/vaadin-select/blob/81e0c22fe858d702b790e67cd08eb3a5b31f73d1/src/vaadin-select.html#L610)
- [vaadin-checkbox](https://github.com/vaadin/vaadin-checkbox/blob/0b3b8e803f6e30c7992d76b0c05be740850e5172/src/vaadin-checkbox.html#L313)
- [vaadin-time-picker](https://github.com/vaadin/vaadin-time-picker/blob/e293b822cd148e23d1116a3d1278027bce4c1b9f/src/vaadin-time-picker.html#L424)
- [vaadin-combo-box](https://github.com/vaadin/vaadin-combo-box/blob/c4be7b3ee458f3ddee89b4eba1cee84cbcb8a756/src/vaadin-combo-box-mixin.html#L759)

Assuming that no one has reported this inconsistency, I think we can drop `sourceEvent`.

With TypeScript, dispatching `change` as `CustomEvent` with `detail.sourceEvent` is in fact rather confusing: people probably expect it to be an `Event` as for native `<input>`.